### PR TITLE
TestCatalogApi implementation

### PR DESCRIPTION
**Pull request by rck109d**

### CODE REVIEW

File: test/test_catalog_api.py [hunk 1]
```diff
@@ -19,68 +19,566 @@
...
+    ci.category_id = '#Beverages'
+    ci.modifier_list_info = [cimli]
+    ci.tax_ids = ['#SalesTax']
+    ci.variations = [st, lt]
+
+    c.item_data = ci
+    return c
+
+
+def build_sales_tax():
```
Comment (frojasg) [inline test/test_catalog_api.py]: could you move these methods inside the test class?

File: test/test_catalog_api.py [hunk 2]
```diff
@@ -19,68 +19,566 @@
...
+    return co
+
+
+class TestCatalogApi(APITestCase):
     """ CatalogApi unit test stubs """
+    objects_by_client_id = dict()
+    test_objects = []
+
+    def build_test_catalog(self):
+        print("build_test_catalog")
```
Comment (frojasg) [inline test/test_catalog_api.py]: Could we remove the print?

File: test/test_catalog_api.py [hunk 3]
```diff
@@ -19,68 +19,566 @@
...
+        while True:
+            res = self.api.list_catalog()
+            if res.objects is None:
+                break
+            ids = set()
+            self.assertTrue(res.errors is None)
+            for co in res.objects:
+                ids.add(co.id)
+                if co.item_data and co.item_data.variations:
+                    for v in co.item_data.variations:
```
Comment (frojasg) [inline test/test_catalog_api.py]: Do we need to iterate over the variations? I think they get listed in the `list_catalog` anyways

File: test/test_catalog_api.py [hunk 4]
```diff
@@ -19,68 +19,566 @@
...
+        ids = self.objects_by_client_id
+        coffee_id = ids['#Coffee']
+        small_coffee_id = ids['#SmallCoffee']
+        large_coffee_id = ids['#LargeCoffee']
+        small_tea_id = ids['#SmallTea']
+
+        delete_request = BatchDeleteCatalogObjectsRequest()
+        delete_request.object_ids = [coffee_id, small_tea_id]
+        response = self.api.batch_delete_catalog_objects(delete_request)
+        self.assertTrue(response.errors is None)
```
Comment (frojasg) [inline test/test_catalog_api.py]: you could use `self.assertIsNone`

File: test/test_catalog_api.py [hunk 5]
```diff
@@ -19,68 +19,566 @@
...
+        self.assertEqual(4, len(response.deleted_object_ids))
+        self.assertTrue(coffee_id in response.deleted_object_ids)
+        self.assertTrue(small_coffee_id in response.deleted_object_ids)
+        self.assertTrue(large_coffee_id in response.deleted_object_ids)
+        self.assertTrue(small_tea_id in response.deleted_object_ids)
```
Comment (frojasg) [inline test/test_catalog_api.py]: you could use `self.assertIn` for these cases

File: test/test_catalog_api.py [hunk 6]
```diff
@@ -19,68 +19,566 @@
...
+    t.percentage = '5.0'
+    t.applies_to_custom_amounts = True
+    t.enabled = True
+
+    co.tax_data = t
+    return co
+
+
+class TestCatalogApi(APITestCase):
     """ CatalogApi unit test stubs """
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove this line?

File: test/test_catalog_api.py [hunk 7]
```diff
@@ -19,68 +19,566 @@
...
+        print("delete_test_catalog")
+        while True:
+            res = self.api.list_catalog()
+            if res.objects is None:
+                break
+            ids = set()
+            self.assertTrue(res.errors is None)
+            for co in res.objects:
+                ids.add(co.id)
+                if co.item_data and co.item_data.variations:
```
Comment (danrice-square) [inline test/test_catalog_api.py]: I don't think you have to call this out, the variations will be deleted when their parent object is deleted.

File: test/test_catalog_api.py [hunk 8]
```diff
@@ -19,68 +19,566 @@
...
+        batch = CatalogObjectBatch(self.test_objects)
+        batches = [batch]
+        req = BatchUpsertCatalogObjectsRequest(str(uuid.uuid4()), batches)
+        res = self.api.batch_upsert_catalog_objects(req)
+        self.assertTrue(res.errors is None)
+        for m in res.id_mappings:
+            self.objects_by_client_id[m.client_object_id] = m.object_id
+
+    def delete_test_catalog(self):
+        print("delete_test_catalog")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 9]
```diff
@@ -19,68 +19,566 @@
...
         self.api = squareconnect.apis.catalog_api.CatalogApi()
+        self.delete_test_catalog()
+        self.build_test_catalog()
 
     def tearDown(self):
-        pass
+        self.delete_test_catalog()
 
     def test_batch_delete_catalog_objects(self):
         print("Start test case for batch_delete_catalog_objects")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 10]
```diff
@@ -19,68 +19,566 @@
...
+        response = self.api.batch_delete_catalog_objects(delete_request)
+        self.assertTrue(response.errors is None)
+        self.assertEqual(4, len(response.deleted_object_ids))
+        self.assertTrue(coffee_id in response.deleted_object_ids)
+        self.assertTrue(small_coffee_id in response.deleted_object_ids)
+        self.assertTrue(large_coffee_id in response.deleted_object_ids)
+        self.assertTrue(small_tea_id in response.deleted_object_ids)
 
     def test_batch_retrieve_catalog_objects(self):
         print("Start test case for batch_retrieve_catalog_objects")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 11]
```diff
@@ -19,68 +19,566 @@
...
+        self.assertTrue(coffee.modifier_list_data is None)
+        self.assertTrue(coffee.modifier_data is None)
+
+        sales_tax = response.objects[1]
+        self.assertEqual('TAX', sales_tax.type)
+        self.assertEqual(sales_tax_id, sales_tax.id)
+        self.assertEqual('Sales Tax', sales_tax.tax_data.name)
 
     def test_batch_upsert_catalog_objects(self):
         print("Start test case for batch_upsert_catalog_objects")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 12]
```diff
@@ -19,68 +19,566 @@
...
+                num_objects += 1
+
+        req = BatchUpsertCatalogObjectsRequest(str(uuid.uuid4()), batches)
+        res = self.api.batch_upsert_catalog_objects(req)
+        self.assertTrue(res.errors is None)
+
+        self.assertEquals(num_objects, len(res.objects))
 
     def test_catalog_info(self):
         print("Start test case for catalog_info")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 13]
```diff
@@ -19,68 +19,566 @@
...
+        self.assertEqual(1000, limits.search_max_page_limit)
+        self.assertEqual(1000, limits.update_item_modifier_lists_max_item_ids)
+        self.assertEqual(1000, limits.update_item_modifier_lists_max_modifier_lists_to_disable)
+        self.assertEqual(1000, limits.update_item_modifier_lists_max_modifier_lists_to_enable)
+        self.assertEqual(1000, limits.update_item_taxes_max_item_ids)
+        self.assertEqual(1000, limits.update_item_taxes_max_taxes_to_disable)
+        self.assertEqual(1000, limits.update_item_taxes_max_taxes_to_enable)
 
     def test_delete_catalog_object(self):
         print("Start test case for delete_catalog_object")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 14]
```diff
@@ -19,68 +19,566 @@
...
+        large_coffee_id = ids['#LargeCoffee']
+        res = self.api.delete_catalog_object(coffee_id)
+
+        self.assertEquals(3, len(res.deleted_object_ids))
+        self.assertTrue(coffee_id in res.deleted_object_ids)
+        self.assertTrue(small_coffee_id in res.deleted_object_ids)
+        self.assertTrue(large_coffee_id in res.deleted_object_ids)
 
     def test_list_catalog(self):
         print("Start test case for list_catalog")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 15]
```diff
@@ -19,68 +19,566 @@
...
-        pass
+        res = self.api.list_catalog()
+        self.assertEqual(len(self.test_objects), len(res.objects))
 
     def test_retrieve_catalog_object(self):
         print("Start test case for retrieve_catalog_object")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 16]
```diff
@@ -19,68 +19,566 @@
...
+            elif obj.type == 'CATEGORY' \
+                    and obj.category_data.name == 'Beverages':
+                got_beverages = True
+
+        self.assertTrue(got_milks)
+        self.assertTrue(got_sales_tax)
+        self.assertTrue(got_beverages)
 
     def test_search_catalog_objects(self):
         print("Start test case for search_catalog_objects")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 17]
```diff
@@ -19,68 +19,566 @@
...
+        for obj in res2.objects:
+            if obj.item_data.name == 'Coffee':
+                got_coffee = True
+            elif obj.item_data.name == 'Tea':
+                got_tea = True
+        self.assertTrue(got_coffee)
+        self.assertTrue(got_tea)
 
     def test_update_item_modifier_lists(self):
         print("Start test case for update_item_modifier_lists")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 18]
```diff
@@ -19,68 +19,566 @@
...
+        res_update = self.api.update_item_taxes(req)
+        self.assertTrue(res_update.errors is None)
+
+        res2 = self.api.retrieve_catalog_object(coffee_id,
+                                                include_related_objects=False)
+        self.assertTrue(res2.errors is None)
+        self.assertTrue(res2.object.item_data.tax_ids is None)
 
     def test_upsert_catalog_object(self):
         print("Start test case for upsert_catalog_object")
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Remove print

File: test/test_catalog_api.py [hunk 19]
```diff
@@ -19,68 +19,566 @@
...
+    co.category_data = cat
+    return co
+
+
+def build_milks():
+    whmm = CatalogModifier()
+    whmm.name = 'Whole Milk'
+    whm = CatalogObject()
+    whm.type = 'MODIFIER'
+    whm.id = '#WholeMilk'
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Maybe define constants for all the '#'-prefixed IDs?

File: test/test_catalog_api.py [hunk 20]
```diff
@@ -19,68 +19,566 @@
...
+def build_beverages():
+    cat = CatalogCategory()
+    cat.name = 'Beverages'
+    co = CatalogObject()
+    co.type = 'CATEGORY'
+    co.id = '#Beverages'
+    co.category_data = cat
+    return co
+
+
```
Comment (danrice-square) [inline test/test_catalog_api.py]: Do we need to have two blank lines every time?

File: test/test_catalog_api.py (hunk 18 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 17 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 16 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 15 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 14 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 13 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 12 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 11 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 10 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 9 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 19 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 20 above)
Comment (rck109d) [inline test/test_catalog_api.py]: resolved after moving method definitions inside of class TestCatalogApi

File: test/test_catalog_api.py (hunk 8 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 5 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 4 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 3 above)
Comment (rck109d) [inline test/test_catalog_api.py]: you are correct; removed variation iteration

File: test/test_catalog_api.py (hunk 2 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 1 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

File: test/test_catalog_api.py (hunk 7 above)
Comment (rck109d) [inline test/test_catalog_api.py]: you are correct; removed variation iteration

File: test/test_catalog_api.py (hunk 6 above)
Comment (rck109d) [inline test/test_catalog_api.py]: done

**Comment by rck109d** [comment]
i addressed all feedback up to now

**Comment by danrice-square** [comment]
LGTM

**Pull request merged**